docs/src/proposals/leader-qos.md

```diff
@@ -0,0 +1,127 @@
+---
+title: Leader QoS
+---
+
+## Motivation:
+Banking thread has too many packets in its buffer, it needs an ability to pack
+high-priority transactions into limited block space first.
+
+## Scope:
+Leader has multiple banking threads, each thread does:
+1. Receives packet_batches from verified_receiver;
+2. Push packet_batches into FIFO buffer;
+3. Traverse FIFO buffer linearly to select packets for processing (aka
+   consuming) or forwarding; Where:
+   
+   "consuming" is a process of preparing selected packets into batches of 
+   sanitized_transactions, submitting for attempted bank execution, which in 
+   turn includes block cost tracking, account locking, and executing/committing.
+   Retryable transactions will be put back to buffer for next iteration.
+
+   "forwarding" is a process of preparing selected unprocessed packets from
+   unforwarded packet_batches, batch-send to Fwd ports. 
+
+The aim of this proposal is to enhance the packet selection process for
+"consuming" and "forwarding" within each banking_stage threads; by 
+compartmentalizing hcanges to speed up delivery. A more holistic change 
+proposal (#23438) exists. 
+
+It propose to replace linear FIFO buffer traversing with transaction-based 
+prioritization. The change scope is mainly at step #3, instead of linearly
+selecting unprocessed packets from buffered packet_batches, it sorts 
+unprocessed packets with following criteria to prioritize consuming and
+forwarding. 
+
+The "consuming" and "forwarding" processing as described above largely remains
+the same, as they are outside scope of this proposal. 
+
+## Prioritization criteria:
+1. Transaction's fee-per-CU (`= (additional_fee + base_fee)/requested_CU)`), 
+   higher paid transactions have priority to be first considered for block 
+   inclusion or forwarding. 
+   Note: `additional_fee` and `requested_cu` are from `compute_budget` i
+   instruction.
+
+2. When fee-per-CU is same, Transactions from higher staked validators have
+   higher chance to be considered for blocking inclusion or forwarding.
+
+## Implementation 
+Because received packets are buffered as packet_batches, in order to 
+prioritize transaction-by-transaction without additional copying of packets,
+there are few extra required changes:
+1. Deserialize packets into transactions upfront, before buffer traversing;
+2. Introduce a way (eg., locator) of iterating packets from buffered 
+   packet_batches;
```

Comment:
not directly related to this change, but why continue to use PacketBatch as opposed to splitting it apart into a deque + trying to build thicc batches of 128 by traversing all pending packets until the end is reached or 128 are locked?

something to do with memory management wrt shifting a bunch of elements?

Reply:
There prolly have few advantages of using PacketBatch, one of them is to support sigverify at gpu. @sakridge can shed more light on it.

@carllin is exploring the very idea (breaking packet_batch into packets within banking_stage) atm.

Reply:
makes sense at sigverify, talking more about after sigverify in banking stage

Reply:
@buffalu I have a draft with these changes here: https://github.com/solana-labs/solana/pull/24187 if you want to take a look

Reply:
oh lets fucking go!

Reply:
Yeah I've run it through some of our automated perf test and the local banking benches, and performance seems in line with master. 

Feel free to suggest/run some other tests you deem interesting